### PR TITLE
feat(pyStoned): Implement data checking

### PR DESCRIPTION
**Pull request by Fangop**
This draft pr provides for pyStoNED.
Both basic models and directional distance function based models are included.
However, it is tricky to test all the circumstance of input.
Hence this is just a draft pr.

Please help me check if the message are providing clear information.

I'm on the work for testing all the models, and trying to adapt `DEA` and `FDH` models to these checking.
So, please do not merge this pr yet.

### CODE REVIEW

**Review by ds2010 — APPROVED** [review]

File: pystoned/utils/tools.py [hunk 1]
```diff
@@ -50,6 +51,96 @@ def trans_list(li):
...
+
+    y_shape = asarray(y).shape
+    x_shape = asarray(x).shape
+
+    if y_shape[0] != x_shape[0]:
+        raise ValueError(
+            "The number of dmu from output data(y) is not equal to the number of dmu from input data(x).")
+
+    if y_shape[1] != len(gy):
+        raise ValueError("The vector of output does not fit the data")
```
Comment (ds2010) [inline pystoned/utils/tools.py:124]: could be:

"Number of outputs must be the same in y and gy."

File: pystoned/utils/tools.py [hunk 2]
```diff
@@ -50,6 +51,96 @@ def trans_list(li):
...
+
+    if x_shape[1] != len(gx):
+        raise ValueError("The vector of input does not fit the data")
```
Comment (ds2010) [inline pystoned/utils/tools.py:127]: "Number of inputs must be the same in X and gx."

File: pystoned/utils/tools.py [hunk 3]
```diff
@@ -50,6 +51,96 @@ def trans_list(li):
...
+        b = trans_list(b)
+        b = to_2d_list(b)
+        gb = to_1d_list(gb)
+        b_shape = asarray(b).shape
+        if b_shape[0] != b_shape[0]:
+            raise ValueError(
+                "The number of dmu from output data(y) is not equal to the number of dmu from undesirable output(b).")
+        if b_shape[1] != len(gb):
+            raise ValueError(
+                "The vector of undesirable output does not fit the data")
```
Comment (ds2010) [inline pystoned/utils/tools.py:139]: "Number of undesirable outputs must be the same in b and gb."

**Review by ds2010 — CHANGES_REQUESTED** [review]

**Comment by ds2010** [comment]
Thanks for your new pr. Here are some suggestions. Please review them again.

**Comment by Fangop** [comment]
@ds2010 Thanks for your comment.
Here are some code for checking the implementation
## CNLS
```python3
# import packages
from pystoned import CNLS
from pystoned.constant import CET_ADDI, FUN_PROD, OPT_LOCAL, RTS_VRS
from pystoned.dataset import load_Finnish_electricity_firm

# import Finnish electricity distribution firms data
data = load_Finnish_electricity_firm(x_select=['Energy', 'Length', 'Customers'],
                                      y_select=['TOTEX'])

# define and solve the CNLS model
model = CNLS.CNLS(y=data.y[1:-1], x=data.x, z=None,
                    cet = CET_ADDI, fun = FUN_PROD, rts = RTS_VRS)
```
This should failed with `ValueError: Number of DMUs must be the same in x and y.`

## CNLS+Z
```python3


# import packages
from pystoned import CNLS
from pystoned.constant import CET_MULT, FUN_COST, RTS_CRS, RED_MOM, OPT_LOCAL
from pystoned.dataset import load_Finnish_electricity_firm

# import all data (including the contextual varibale)
data = load_Finnish_electricity_firm(x_select=['Energy', 'Length', 'Customers'],
                                     y_select=['TOTEX'],
                                     z_select=['PerUndGr'])

# define and solve the CNLS-Z model
model = CNLS.CNLS(y=data.y, x=data.x, z=data.z[1:-1], cet = CET_MULT, fun = FUN_COST, rts = RTS_CRS)
model.optimize("test@gmail.com")

# display the coefficient of contextual variable
model.display_lamda()
```
This should failed with `ValueError: Number of DMUs must be the same in y and z`.

## CNLSG
```
# import packages
from pystoned import CNLSG
from pystoned.constant import CET_ADDI, FUN_PROD, OPT_LOCAL, RTS_VRS
from pystoned.dataset import load_Finnish_electricity_firm

# import Finnish electricity distribution firms data
data = load_Finnish_electricity_firm(x_select=['Energy', 'Length', 'Customers'],
                                      y_select=['TOTEX'])

# define and solve the CNLS model
model = CNLSG.CNLSG(y=data.y[1:-1], x=data.x, z=None,
                    cet = CET_ADDI, fun = FUN_PROD, rts = RTS_VRS)
```

This should also failed with `ValueError: Number of DMUs must be the same in x and y.`

## CNLSDDF
```python3
# import packages
from pystoned import CNLSDDF
from pystoned.constant import FUN_PROD, OPT_LOCAL
from pystoned.dataset import load_Finnish_electricity_firm

# import Finnish electricity distribution firms data
data = load_Finnish_electricity_firm(x_select=['OPEX', 'CAPEX'],
                                    y_select=['Energy', 'Length', 'Customers'])

# define and solve the CNLS-DDF model
model = CNLSDDF.CNLSDDF(y=data.y, x=data.x[1:-1], b=None, fun = FUN_PROD, gx= [1.0, 0.0], gb=None, gy= [0.0, 0.0, 0.0])
model.optimize(OPT_LOCAL)
```
This should failed with `ValueError: Number of DMUs must be the same in x and y.`.

## DEA
```python3
# import packages
from pystoned import DEA
from pystoned import dataset as dataset
from pystoned.constant import RTS_VRS, ORIENT_IO, OPT_LOCAL

# import the data provided with Tim Coelli’s Frontier 4.1
data = dataset.load_Tim_Coelli_frontier()

# define and solve the DEA radial model
model = DEA.DEA(data.y[1:-1], data.x, rts=RTS_VRS, orient=ORIENT_IO, yref=None, xref=None)
model.optimize(OPT_LOCAL)
```
This should failed with `ValueError: Number of DMUs must be the same in x and y.`

All the other models are equipped with data checking. Please help me check if there are some left.

The pr is  ready for merge now, thank you!

**Comment by ds2010** [comment]
@Fangop, it seems that `pyomo.core.base.plugin` module is deprecated in pyomo. See the following warning (run above examples). Should we need to worry about this? Is it necessary to update the deprecated module used in pyStoned?

``` Python3
WARNING: DEPRECATED: The pyomo.core.base.plugin module is deprecated.  See
    pyomo.core.base.transformation for Transformation and
    TransformationFactory, pyomo.core.base.component for ModelComponentFactory
    and pyomo.scripting.interface for IPyomoScript* interfaces.  (deprecated
    in 6.0) (called from <frozen importlib._bootstrap>:219)
c:/Users/dais2/Downloads/test.py:7: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.
```

**Pull request merged**